app/src/main/java/tttTDD/Java/Game.java

```diff
@@ -67,11 +67,21 @@ public boolean performMove(int playerInput) {
         return true;
     }
 
-    private boolean gameHasWinner() {
-        if (board.hasWinningSet(currentPlayer.getMarker())) {
+    private boolean gameHasFinished() {
+        if (isThereAWinner() && !isThereATie()) {
```

Comment:
I personally love reading conditionals in human-readable format. Sometimes little conditionals like these might feel like it's overkill or indirection, but the naming and readability counter that and make it very worth our time. I also like how you did not combine those into a single conditional method that then calls those two. It shows us that there is more than one scenario in the winning check without overloading us. However, I would tack on to that, in times that I've ended up with complex well-named conditional methods, I have had to create a wrapper for them just to make that line readable. It's rare though that I've felt the need to do so.

Reply:
Second concept: Spot glancing the code, I believe you could delete "&& !isThereATie()". I would be curious if you delete that, do all of your tests still pass, and does the code still work? It feels like it's not needed, but not 100% unless I dug in more.

Reply:
This one was a bit funky since it had to encompass the spirit of game completion rather than declaring a winner. By this logic you're also right that it doesn't need `&& !isThereATie` since the message is only for the winner.